src/Servers/Kestrel/Core/src/Internal/Http/HttpRequestPipeReader.cs

```diff
@@ -99,7 +99,10 @@ public void Abort(Exception error = null)
             if (_state != HttpStreamState.Closed)
             {
                 _state = HttpStreamState.Aborted;
-                _error = error;
+                if (error != null)
+                {
+                    _error = ExceptionDispatchInfo.Capture(error);
```

Comment:
I approved it since it feels more standard/correct, but is there an actual functional impact to this?

For example, if the same exception instance gets rethrown somewhere else after we "capture" it here, will it save us from those super-long stack traces being built up from the multiple throwers of the same exception?

Reply:
The exception is mutable; so each `.Capture(error)` adds another stack to the original exception; combined with the `.Capture(error).Throw()` it means each time `ValidateState` is called and throws it will add anew stack to the exception so each Throw will get a longer and longer stack trace (if it happens multiple times with the same exception).

Changing this to only capture once; means even if it throws multiple times the stack won't grow and grow (concurrent throws aside).

>  Was it the multiple calls to ExceptionDispatchInfo.Capture() itself that created the long stack traces, not the throwing of the exception multiple times?

This would do that; however... I don't see `HttpRequestPipeReader` in the stacks, so I'm not sure its the cause?

Reply:
Putting it another way, this is a definite issue with what its doing here and a fix

Not sure about the other though; perhaps if the same exception is getting Edi captured in two or more places https://github.com/aspnet/AspNetCore/pull/11827 then the stack could get wacky?

Reply:
e.g.
```csharp

interface IThrow { void Throw(ExceptionDispatchInfo edi, Exception ex); }

struct EdiThrow : IThrow
{
    public void Throw(ExceptionDispatchInfo edi, Exception ex)
    {
        edi.Throw();
    }
}

struct EdiCaptureThrow : IThrow
{
    public void Throw(ExceptionDispatchInfo edi, Exception ex)
    {
        ExceptionDispatchInfo.Capture(ex).Throw();
    }
}

class Program
{
    static ExceptionDispatchInfo s_edi;
    static Exception s_e;

    static void Main(string[] args)
    {
        Exception exception = null;

        SetException();
        for (var i = 0; i < 4; i++)
        {
            exception = Throw(new EdiThrow());
        }

        Console.WriteLine("EdiThrow");
        Console.WriteLine(exception.StackTrace);
        Console.WriteLine();

        SetException();
        for (var i = 0; i < 4; i++)
        {
            exception = Throw(new EdiCaptureThrow());
        }

        Console.WriteLine("EdiCaptureThrow");
        Console.WriteLine(exception.StackTrace);
    }



    static void SetException()
    {
        try
        {
            throw new Exception("source");
        }
        catch (Exception ex)
        {
            s_e = ex;
            s_edi = ExceptionDispatchInfo.Capture(ex);
        }
    }

    static Exception Throw<TThrow>(TThrow thrower) where TThrow : IThrow
    {
        Exception exception = null;
        try { thrower.Throw(s_edi, s_e); }
        catch (Exception ex) { exception = ex; }
        return exception;
    }
}
```
```
EdiThrow
   at Program.SetException() in C:\Work\test\Exceptions\Program.cs:line 60
--- End of stack trace from previous location where exception was thrown ---
   at EdiThrow.Throw(ExceptionDispatchInfo edi, Exception ex) in C:\Work\test\Exceptions\Program.cs:line 13
   at Program.Throw[TThrow](TThrow thrower) in C:\Work\test\Exceptions\Program.cs:line 72

EdiCaptureThrow
   at Program.SetException() in C:\Work\test\Exceptions\Program.cs:line 60
--- End of stack trace from previous location where exception was thrown ---
   at EdiCaptureThrow.Throw(ExceptionDispatchInfo edi, Exception ex) in C:\Work\test\Exceptions\Program.cs:line 21
   at Program.Throw[TThrow](TThrow thrower) in C:\Work\test\Exceptions\Program.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at EdiCaptureThrow.Throw(ExceptionDispatchInfo edi, Exception ex) in C:\Work\test\Exceptions\Program.cs:line 21
   at Program.Throw[TThrow](TThrow thrower) in C:\Work\test\Exceptions\Program.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at EdiCaptureThrow.Throw(ExceptionDispatchInfo edi, Exception ex) in C:\Work\test\Exceptions\Program.cs:line 21
   at Program.Throw[TThrow](TThrow thrower) in C:\Work\test\Exceptions\Program.cs:line 72
--- End of stack trace from previous location where exception was thrown ---
   at EdiCaptureThrow.Throw(ExceptionDispatchInfo edi, Exception ex) in C:\Work\test\Exceptions\Program.cs:line 21
   at Program.Throw[TThrow](TThrow thrower) in C:\Work\test\Exceptions\Program.cs:line 72
```